ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -0,0 +1,145 @@
+};
+
+static void cpu_and_wall_time_collector_typed_data_mark(void *state_ptr) {
+  struct cpu_and_wall_time_collector_state *state = (struct cpu_and_wall_time_collector_state *) state_ptr;
+
+  // Update this when modifying state struct
+  rb_gc_mark(state->recorder_instance);
+}
+
+static void cpu_and_wall_time_collector_typed_data_free(void *state_ptr) {
+  struct cpu_and_wall_time_collector_state *state = (struct cpu_and_wall_time_collector_state *) state_ptr;
+
+  // Update this when modifying state struct
+
+  // Important: Remember that we're only guaranteed to see here what's been set in _native_new, aka
+  // pointers that have been set NULL there may still be NULL here.
+  if (state->sampling_buffer != NULL) sampling_buffer_free(state->sampling_buffer);
+
+  xfree(state);
+}
+
+static VALUE _native_new(VALUE klass) {
+  struct cpu_and_wall_time_collector_state *state = xcalloc(1, sizeof(struct cpu_and_wall_time_collector_state));
+
+  if (state == NULL) {
+    rb_raise(rb_eNoMemError, "Failed to allocate memory for components of Datadog::Profiling::Collectors::CpuAndWallTime");
+  }
+
+  // Update this when modifying state struct
+  state->sampling_buffer = NULL;
+  state->recorder_instance = Qnil;
+
+  return TypedData_Wrap_Struct(collectors_cpu_and_wall_time_class, &cpu_and_wall_time_collector_typed_data, state);
+}
+
+static VALUE _native_initialize(VALUE self, VALUE collector_instance, VALUE recorder_instance, VALUE max_frames) {
+  enforce_recorder_instance(recorder_instance);
+
+  struct cpu_and_wall_time_collector_state *state;
+  TypedData_Get_Struct(collector_instance, struct cpu_and_wall_time_collector_state, &cpu_and_wall_time_collector_typed_data, state);
+
+  int max_frames_requested = NUM2INT(max_frames);
+  if (max_frames_requested < 0) rb_raise(rb_eArgError, "Invalid max_frames: value must not be negative");
+
+  // Update this when modifying state struct
+  state->sampling_buffer = sampling_buffer_new(max_frames_requested);
+  state->recorder_instance = recorder_instance;
+
+  return Qtrue;
+}
+
+// This method exists only to enable testing Datadog::Profiling::Collectors::CpuAndWallTime behavior using RSpec.
+// It SHOULD NOT be used for other purposes.
+static VALUE _native_sample(VALUE self, VALUE collector_instance) {
+  sample(collector_instance);
+  return Qtrue;
+}
+
+static void sample(VALUE collector_instance) {
+  struct cpu_and_wall_time_collector_state *state;
+  TypedData_Get_Struct(collector_instance, struct cpu_and_wall_time_collector_state, &cpu_and_wall_time_collector_typed_data, state);
+
+  VALUE threads = ddtrace_thread_list();
+
+  const long thread_count = RARRAY_LEN(threads);
+  for (long i = 0; i < thread_count; i++) {
+    VALUE thread = RARRAY_AREF(threads, i);
+
+    int64_t metric_values[ENABLED_VALUE_TYPES_COUNT] = {0};
+
+    // FIXME: TODO
+    metric_values[CPU_TIME_VALUE_POS] = 12;
+    metric_values[CPU_SAMPLES_VALUE_POS] = 34;
+    metric_values[WALL_TIME_VALUE_POS] = 56;
```

Comment:
Is there more information about these "magic" numbers?

Reply:
I've clarified the comment to highlight that they are just dummy values: https://github.com/DataDog/dd-trace-rb/pull/2019/commits/f39ab914fe9e824f4cfedee173f14bd43e8f9aaa .

The intention here is that I'll store a per-thread structure which tracks the last seen values for cpu and wall time and calculates the time elapsed since last sampled. This is the same we do in the current Ruby profiler, see https://github.com/DataDog/dd-trace-rb/blob/b8801d15bc25ab64abee941ef61fc2457104f1f0/lib/datadog/profiling/collectors/stack.rb#L152-L154.